crates/synthesis/src/two_qubit_decompose.rs

```diff
@@ -2531,66 +2545,38 @@ impl TwoQubitControlledUDecomposer {
-                    let inv_gate = gate
-                        .inverse(&params.into_iter().map(Param::Float).collect::<Vec<_>>())
-                        .unwrap();
-                    let inv_gate_params = inv_gate
-                        .1
-                        .into_iter()
-                        .map(|param| match param {
-                            Param::Float(val) => val,
-                            _ => unreachable!(
-                                "Parameterized inverse generated from non-parameterized gate."
-                            ),
-                        })
-                        .collect::<SmallVec<_>>();
-                    Ok((Some(inv_gate.0), inv_gate_params, qubits))
-                }
-                RXXEquivalent::CustomPython(gate_cls) => Python::with_gil(|py: Python| {
-                    let gate_obj = gate_cls.bind(py).call1(PyTuple::new(py, params)?)?;
-                    let raw_inverse = gate_obj.call_method0(intern!(py, "inverse"))?;
-                    let inverse: OperationFromPython = raw_inverse.extract()?;
-                    let params: SmallVec<[f64; 3]> = inverse
-                        .params
-                        .into_iter()
-                        .map(|x| match x {
-                            Param::Float(val) => val,
-                            _ => panic!("Inverse has invalid parameter"),
-                        })
-                        .collect();
-                    if let Some(gate) = inverse.operation.try_standard_gate() {
-                        Ok((Some(gate), params, qubits))
-                    } else if raw_inverse.is_instance(gate_cls.bind(py))? {
-                        Ok((None, params, qubits))
-                    } else {
-                        Err(QiskitError::new_err(
-                            "rxx gate inverse is not valid for this decomposer",
-                        ))
-                    }
-                }),
+        let inv_gate = match gate.view() {
+            OperationRef::StandardGate(gate) => {
+                let res = gate
+                    .inverse(&params.into_iter().map(Param::Float).collect::<Vec<_>>())
+                    .unwrap();
+                (res.0.into(), res.1)
             }
-        }
+            OperationRef::Gate(gate) => {
+                Python::with_gil(|py: Python| -> PyResult<(PackedOperation, SmallVec<_>)> {
+                    let raw_inverse = gate.gate.call_method0(py, intern!(py, "inverse"))?;
+                    let inverse: OperationFromPython = raw_inverse.extract(py)?;
+                    Ok((inverse.operation, inverse.params))
+                })?
+            }
+            // UnitaryGate isn't applicable here as the 2q gate here is the parameterized
+            // ControlledU equivalent used in the decomposition. This precludes UnitaryGate
+            _ => unreachable!("Only 2q gate objects can be inverted in the decomposer"),
```

Comment:
IMO this is more of a case for `panic!` since it's technically reachable if we call the function with the wrong kind of operations :3

Reply:
It's a private method that will only get called by this struct itself. In the constraints of what is going to call this it is unreachable because the struct can never have a `UnitaryGate` in the output sequence. If it has we've broken something more fundamental. I'm fine switching it to a panic though it doesn't functionally matter they're both panics, `unreachable` just changes the error message slightly.

Reply:
Done in: https://github.com/Qiskit/qiskit/pull/14650/commits/4f1b1bb07a7b8aaf9585da6e2b9664273c7ecacd